### PR TITLE
Calculate average power readings over a time period

### PR DESCRIPTION
**Pull request by jordanterry**
A single `DataEntry` represents a single power reading in time. We want to work on a data entry that represents power over a second.
To do this we need to represent a seconds worth of power as a point of time at the start of a second and a point of time at the end of the second. Currently this is represented as a `SecondPower`.

| Power Analysis Tool | Training Peaks |
| --- | --- |
| Screenshot 2020-07-05 at 11 59 21 | Screenshot 2020-07-05 at 11 59 52 |
| | Screenshot 2020-07-05 at 12 00 14 |
| | Screenshot 2020-07-05 at 12 00 27 |

 |

### CODE REVIEW

File: src/PowerReadingConverter.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+     * 
+     * An average power of one second is calculated by taking the power at the start and end of a second in time and then calculating the 
+     * average. 
+     * 
+     * TODO(jordan): Use the time of a DataEntry to identify a second.
+     *
+     * @param readings data entry readings to convert to second data
+     */
+    public convert(dataEntries: DataEntry[]): AverageSecondPower[] {
+        return dataEntries.reduce(function(second: AverageSecondPower[], currentValue: DataEntry, currentIndex: number, array: DataEntry[]) {
```
Comment (philharper) [inline src/PowerReadingConverter.ts]: Can this be changed to use an arrow function? I think it makes it a little cleaner.

File: src/PowerReadingConverter.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import DataEntry from "./types/DataEntry";
```
Comment (philharper) [inline src/PowerReadingConverter.ts]: Can these new files be nested in a converter directory?

File: src/PowerReadingConverter.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import AverageSecondPower from "./types/AverageSecondPower";
+
+export class PowerReadingConverter {
+   /**
+     * Convert an array of DataEntry to an array contaning the average power of one second.
+     * 
+     * An average power of one second is calculated by taking the power at the start and end of a second in time and then calculating the 
+     * average. 
+     * 
+     * TODO(jordan): Use the time of a DataEntry to identify a second.
```
Comment (philharper) [inline src/PowerReadingConverter.ts]: I think could be done in the gpx converter.

File: src/types/AverageRange.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+    private readonly averages: AverageSecondPower[]
+    readonly range: number
+
+    constructor(averages: AverageSecondPower[], range: number) {
+        this.averages = averages
+        this.range = range
+    }
+
+    average() {
```
Comment (jordanterry) [inline src/types/AverageRange.ts]: Need to add return type.

File: src/types/AverageSecondPower.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+export default class AverageSecondPower {
+
+    private readonly start: DataEntry
+    private readonly end: DataEntry
+    readonly average: number
+
+    constructor(start: DataEntry, end: DataEntry) {
+        this.start = start
+        this.end = end
+        this.average = ((this.start.power1 || 0) + (this.end.power1 || 0)) / 2
```
Comment (jordanterry) [inline src/types/AverageSecondPower.ts]: Use `number` rather than `DataEntry` to remove nullability check

File: src/utils/PowerAnalysisUtils.ts [hunk 6]
```diff
@@ -25,32 +28,14 @@ export default class PowerAnalysisUtils {
         return Math.floor(totalPower / powerReadings.length);
     }
 
-    static calculateMaxAveragePowerByTime(powerReadings: DataEntry[], seconds: number) {
+    static calculateMaxAveragePowerByTime(powerReadings: AverageSecondPower[], seconds: number) {
```
Comment (jordanterry) [inline src/utils/PowerAnalysisUtils.ts:31]: An immutable array would be better than an array which can be manipulated by accident.

File: src/utils/PowerAnalysisUtils.ts [hunk 7]
```diff
@@ -25,32 +28,14 @@ export default class PowerAnalysisUtils {
...
         let maxAveragePower = 0;
         for (let i = 0; i < powerReadings.length; i++) {
-            let rangeTotalPower = this.getRangeTotalPower(i, seconds, powerReadings);
-
-            const rangeAveragePower = rangeTotalPower / seconds;
+            const averageRange = new AverageRange(powerReadings.splice(i, i + seconds), seconds)
+            const rangeAveragePower = averageRange.average();
```
Comment (jordanterry) [inline src/utils/PowerAnalysisUtils.ts:35]: Maybe write all averages to a list and return the highest value.

File: src/powerchart/PowerAverages.tsx [hunk 8]
```diff
@@ -11,22 +12,24 @@ function PowerAverages(props: PowerAverageProps) {
     const secondPowers = [5, 10, 30];
     const minutePowers = [1, 5, 10, 20];
     const data = props.data;
+    
+    const avg = new PowerReadingConverter().convert(data)
```
Comment (philharper) [inline src/powerchart/PowerAverages.tsx]: Can we rename this to average?

**Review by philharper — COMMENTED** [review]
Happy to merge, are you going to make changes that the comments have raised?

File: src/PowerReadingConverter.ts (hunk 1 above)
Comment (jordanterry) [reply src/PowerReadingConverter.ts]: Changed, didn't realise the arrow function was Typescript and the function was Javascript.

File: src/powerchart/PowerAverages.tsx (hunk 8 above)
Comment (jordanterry) [reply src/powerchart/PowerAverages.tsx]: 👍

**Comment by jordanterry** [comment]
@philharper changes made, if you approve we can merge it.

**Pull request merged**